agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -0,0 +1,106 @@
+require_relative '../service_pods/creator'
+require_relative '../service_pods/stopper'
+require_relative '../service_pods/terminator'
+
+module Kontena::Workers
+  class ServicePodWorker
+    include Celluloid
+    include Kontena::Logging
+    include Kontena::ServicePods::Common
+    include Kontena::Helpers::RpcHelper
+
+    attr_reader :node, :prev_state, :service_pod
+    attr_accessor :service_pod
+
+    def initialize(node, service_pod = nil)
+      @node = node
+      @service_pod = service_pod
+      @prev_state = 'unknown'
+    end
+
+    def update(service_pod = nil)
+      @service_pod = service_pod if service_pod
+      ensure_desired_state
+    end
+
+    def destroy
+      @service_pod.mark_as_terminated
+      update
+    end
+
+    def ensure_desired_state
```

Comment:
The design needs something to report the resulting error states back to the master as part of the service instance state.